hashtable.t

```diff
@@ -0,0 +1,509 @@
+		var calloc_result = table_calloc(initial_capacity)
+		
+		if calloc_result:is_ok() then
+			var opaque_ptr, metadata, buckets = calloc_result.ok
+			
+			self:_init {
+				capacity = initial_capacity,
+				size = 0,
+				opaque_ptr = opaque_ptr,
+				metadata = metadata,
+				buckets = buckets
+			}
+		end
+	end
+
+	terra HashTable:destruct()
+		if self.opaque_ptr ~= nil then
+			[Alloc]:free_raw(self.opaque_ptr)
+			CStr.memset(self, 0, sizeof(HashTable))
+		end
+	end
+
+	-- Resizes the hashtable to be at least the size of the requested_capacity.
+	-- Returns 0 if there was no error. 
+	terra HashTable:reserve(requested_capacity: uint): uint
+		-- If the requested_capacity is lower than self.capacity, return
+		if requested_capacity < self.capacity then
+			return M.Errors.ValueError
+		end
+
+		var new_capacity = find_next_power_of_two(requested_capacity) 
+		var new_opaque, new_metadata, new_buckets = table_calloc(new_capacity):unwrap()
+
+		-- Iterate through the old data and rehash existing entries
+		for i = 0, self.capacity do
+			if metadata_holds_hash(self.metadata[i]) then
+				var bucket = self.buckets[i]
+				var insert_result = insert_bucket(bucket, new_metadata, new_buckets, new_capacity)
+
+				if insert_result:is_err() then
+					[Alloc]:free_raw(new_opaque)
+					new_opaque = nil
+					new_metadata = nil
+					new_buckets = nil
+					return insert_result.err
+				end
+			end
+		end
+
+		[Alloc]:free_raw(self.opaque_ptr)
+		self.opaque_ptr = new_opaque
+		self.metadata = new_metadata
+		self.buckets = new_buckets
+		self.capacity = new_capacity
+
+		return 0
+	end
+
+	terra HashTable:has(key: KeyType): bool
+		var result = hash_probe(key, self.metadata, self.buckets, self.capacity) 
+
+		if result:is_ok() and metadata_holds_hash(self.metadata[result.ok._1]) then
+			return true
+		else
+			return false
+		end
+	end
+
+	-- Removes the item associated with the key and returns the a struct containing a key field and a possible value field.
+	terra HashTable:remove(key: KeyType): RemoveResult
+		var hash_info, index = hash_probe(key, self.metadata, self.buckets, self.capacity):unwrap()
+
+		if metadata_holds_hash(self.metadata[index]) then
+			var bucket = self.buckets[index]
+			
+			self.metadata[index] = MetadataDeleted
+			CStr.memset(self.buckets + index, 0, sizeof(BucketType))
+			self.size = self.size - 1
+			
+			return RemoveResult.ok(bucket)
+		end
+
+		return RemoveResult.err(M.Errors.NotFound)
+	end
+
+	local InsertBody = macro(function(self, bucket)
+		return quote
+			var this = [self]
+			if is_at_capacity(this.capacity, this.size) then
+				-- Reserve computes the next power of two, so this will double capacity.
+				-- passing in [self].capacity * 2 will quadruple capacity instead
+				this:reserve(this.capacity + 1)
+			end
+			
+			var insert_data = insert_bucket([bucket], this.metadata, this.buckets, this.capacity):unwrap()
+
+			if insert_data.old_metadata == MetadataEmpty then
```

Comment:
What about if it is being inserted in place of a deleted element? Wouldn't that increase the size too?